### PR TITLE
[Beta Tester] Add tool to force wccom endpoint errors or timeout 

### LINKED ISSUE #50494: Add beta tester tool to simulate woocommerce.com downtime
**Prerequisites (mark completed items with an [x]):**
- [ ] I have checked that my issue type is not listed here https://github.com/woocommerce/woocommerce/issues/new/choose
- [ ] My issue is not a security issue, support request, bug report, enhancement or feature request (Please use the link above if it is).

**Issue Description:**

To help test the behavior of WooCommerce when WooCommerce.com has downtime, create a tool in beta tester that creates errors fors requests to WooCommerce.com.

See @moon0326's gist for an example  https://gist.github.com/moon0326/f395d07180e9f0cb1704cc503c442727

## Acceptance Criteria

* Create a toggle in beta tester to turn on / off WooCommerce.com simulated downtime.
* Modify the requests to respond with 501 HTTP error or similar
* Add a 10s or similar delay to the response

### PR DESCRIPTION
**Pull request by psealock**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/50494

This PR adds a tool to the Beta Tester to force returned errors on all requests to `woocommerce.com` endpoints to simulate a down or slow server on woocommerce.com.

image

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Its easiest to test this one locally.

1. Tools > WCA Helper > Tools > Force errors on woocommerce.com requests.
2. Choose timeout as the error mode and click "Run".
3. Clear your transients to make sure were not reading from cache `wp transient delete --all`.
4. Go the homescreen and note the slow load time. There are 4 requests to wccom being made on the backend and each will sleep for 6 seconds and result in a timeout error.
5. As of https://github.com/woocommerce/woocommerce/pull/50521 the errors will result in default values being used. Refresh the Homescreen and note normal load times as the cached values are used and a new request is avoided.
6. Now try the Error mode in WCA Test Helper and clear your cache once again `wp transient delete --all`.
7. The homescreen should now load normally, once again with default values as a result of the errors. No errors should be visible on the client.
8. Bonus testing: Add logging to the data poller to confirm errors are being handled. Add some to the Payment Gateway Suggestions as well to see default values being returned as a result.

https://github.com/woocommerce/woocommerce/blob/175ebe2a5e7cc3205ebabc0ef7f10fe02deb69eb/plugins/woocommerce/src/Admin/RemoteSpecs/DataSourcePoller.php#L196

https://github.com/woocommerce/woocommerce/blob/175ebe2a5e7cc3205ebabc0ef7f10fe02deb69eb/plugins/woocommerce/src/Admin/Features/PaymentGatewaySuggestions/Init.php#L74

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Add a tool to simulate server problems with woocommerce.com

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-beta-tester/includes/class-wc-beta-tester-wccom-requests.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        if ( 'error' === $mode ) {
+            add_filter( 'pre_http_request', array( $this, 'override_http_request_error' ), 10, 3 );
+        }
+	}
+
+    /**
+     * Override the http request with a timeout.
+     */
+    public function override_http_request_timeout( $response, $args, $url ) {
+        if ( strpos( $url, 'https://woocommerce.com/wp-json/' ) !== false ) {
```
Comment (ilyasfoo) [inline plugins/woocommerce-beta-tester/includes/class-wc-beta-tester-wccom-requests.php]: Sometimes I manually edit the data source to point to `http://woocommerce.test/`. Could we also catch that URL? 🙏

File: plugins/woocommerce-beta-tester/src/tools/commands/set-wccom-request-errors.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Internal dependencies
+ */
+import { STORE_KEY } from '../data/constants';
+
+export const UPDATE_WCCOM_REQUEST_ERRORS_MODE = 'updateWccomRequestErrorsMode';
+
+const OPTIONS = [
+	{ label: 'Timeout', value: 'timeout' },
```
Comment (ilyasfoo) [inline plugins/woocommerce-beta-tester/src/tools/commands/set-wccom-request-errors.js]: A bit more verbose helps I think
```suggestion
	{ label: 'Timeout after 6 seconds', value: 'timeout' },
```

**Review by ilyasfoo — APPROVED** [review]
Sweet, great work! I caught a couple of [sync calls](https://github.com/woocommerce/woocommerce/issues/50543#issuecomment-2288744885) with this tool.

Nit comment but pre-approving!

File: plugins/woocommerce-beta-tester/includes/class-wc-beta-tester-wccom-requests.php (hunk 1 above)
Comment (psealock) [reply plugins/woocommerce-beta-tester/includes/class-wc-beta-tester-wccom-requests.php]: Nice, updated to check local `woocommerce.test` as well in https://github.com/woocommerce/woocommerce/pull/50654/commits/8d64ad089598588735cf7309ff43d44a1f581448

**Pull request merged**